zerver/worker/user_presence.py

```diff
@@ -22,4 +23,5 @@ def consume(self, event: Mapping[str, Any]) -> None:
         client = get_client(event["client"])
         log_time = timestamp_to_datetime(event["time"])
         status = event["status"]
-        do_update_user_presence(user_profile, client, log_time, status)
+        with transaction.atomic():
+            do_update_user_presence(user_profile, client, log_time, status)
```

Comment:
`do_update_user_presence` seems like something that should definitely be atomic, so not having that is a bit of a bug

Reply:
Hmm, yeah, Might be cleaner to do this via a decorator on the `do_` function, though.